File: ios/LiulianUI/Sources/LiulianUI/LiulianButton.swift
import SwiftUI

/// LiulianButton — per ui-spec/Button.spec.md.
///
/// Renders via SwiftUI primitives, NOT system UIButton. Disables default highlights,
/// renders own press scale + focus ring. Touch target padded to platform minimum.
public struct LiulianButton: View {
    public enum Variant {
        case primary
        case secondary
        case ghost
        case danger
    }

    public enum Size {
        case sm
        case md
        case lg
    }

    let label: String
    let variant: Variant
    let size: Size
    let loading: Bool
    let disabled: Bool
    let onPress: () -> Void

    @State private var isPressed: Bool = false
    @State private var isHovered: Bool = false
    @FocusState private var isFocused: Bool

    public init(
        _ label: String,
        variant: Variant = .primary,
        size: Size = .md,
        loading: Bool = false,
        disabled: Bool = false,
        onPress: @escaping () -> Void
    ) {
        self.label = label
        self.variant = variant
        self.size = size
        self.loading = loading
        self.disabled = disabled
        self.onPress = onPress
    }

    public var body: some View {
        Button(action: { if !loading && !disabled { onPress() } }) {
            HStack(spacing: LiulianTokens.Control.IconGap.value) {
                if loading {
                    // Simple rotating circle as our spinner — matches design-system motion token.
                    ProgressView()
                        .progressViewStyle(.circular)
                        .tint(foregroundColor)
                        .scaleEffect(0.8)
                } else {
                    LiulianText(label, variant: textVariant, color: foregroundColor)
                }
            }
            .frame(height: height)
            .padding(.horizontal, paddingX)
            .background(backgroundColor)
            .clipShape(RoundedRectangle(cornerRadius: LiulianTokens.Radius.md))
            .overlay(borderOverlay)
            .scaleEffect(isPressed ? LiulianTokens.Control.ScaleActive.value : 1.0)
            .offset(y: isHovered && !isPressed ? -1 : 0)
            .opacity(disabled ? LiulianTokens.Opacity.disabled : 1.0)
            // Focus ring (always visible when focused, per spec)
            .overlay(
                RoundedRectangle(cornerRadius: LiulianTokens.Radius.md)
                    .strokeBorder(
                        isFocused ? LiulianTokens.Colors.unibeRed : Color.clear,
                        lineWidth: LiulianTokens.Focus.ringWidth
                    )
                    .padding(-(LiulianTokens.Focus.ringOffset))
            )
            .animation(
                .timingCurve(0.16, 1.0, 0.3, 1.0, duration: LiulianTokens.Duration.fast),
                value: isHovered
            )
            .animation(
                .timingCurve(0.22, 1.0, 0.36, 1.0, duration: LiulianTokens.Duration.instant),
                value: isPressed
            )
        }
        .buttonStyle(.plain)
        .focused($isFocused)
        .disabled(disabled || loading)
        .accessibilityLabel(Text(label))
        .accessibilityAddTraits(.isButton)
        .accessibilityHint(loading ? Text("Loading") : Text(""))
        // Ensure tappable area meets platform min target
        .frame(minHeight: LiulianTokens.Touch.minTargetIos)
        .contentShape(Rectangle())
        .onLongPressGesture(
            minimumDuration: .infinity,
            maximumDistance: .infinity,
            perform: {},
            onPressingChanged: { pressing in isPressed = pressing }
        )
#if os(macOS)
        .onHover { hovering in isHovered = hovering }
#endif
    }

    // MARK: - Style resolution

    private var height: CGFloat {
        switch size {
        case .sm: return LiulianTokens.Control.Height.sm
        case .md: return LiulianTokens.Control.Height.md
        case .lg: return LiulianTokens.Control.Height.lg
        }
    }

    private var paddingX: CGFloat {
        switch size {
        case .sm: return LiulianTokens.Control.PaddingX.sm
        case .md: return LiulianTokens.Control.PaddingX.md
        case .lg: return LiulianTokens.Control.PaddingX.lg
        }
    }

    private var textVariant: LiulianText.Variant {
        switch size {
        case .sm: return .caption
        case .md: return .body
        case .lg: return .subtitle
        }
    }

    private var backgroundColor: Color {
        let base: Color
        switch variant {
        case .primary: base = LiulianTokens.Colors.unibeRed
        case .secondary: base = LiulianTokens.Colors.surfacePure
        case .ghost:    base = Color.clear
        case .danger:   base = LiulianTokens.Colors.unibeRedDeep
        }
        if isPressed { return base.opacity(0.95) }
        if isHovered {
            switch variant {
            case .primary: return LiulianTokens.Colors.unibeRedDeep
            case .secondary: return LiulianTokens.Colors.canvasWarm
            case .ghost:   return LiulianTokens.Colors.canvasWarm
            case .danger:  return LiulianTokens.Colors.unibeRedDeep
            }
        }
        return base
    }

    private var foregroundColor: Color {
        switch variant {
        case .primary, .danger: return LiulianTokens.Colors.surfacePure
        case .secondary:        return LiulianTokens.Colors.inkCharcoal
        case .ghost:            return LiulianTokens.Colors.inkMuted
        }
    }

    @ViewBuilder
    private var borderOverlay: some View {
        if variant == .secondary {
            RoundedRectangle(cornerRadius: LiulianTokens.Radius.md)
                .strokeBorder(
                    isHovered ? LiulianTokens.Colors.hairlineStrong : LiulianTokens.Colors.hairline,
                    lineWidth: 1
                )
        }
    }
}

// MARK: - Preview

#if DEBUG
struct LiulianButton_Previews: PreviewProvider {
    static var previews: some View {
        VStack(alignment: .leading, spacing: LiulianTokens.Spacing.s4) {
            LiulianText("Variants", variant: .monoLabel, color: LiulianTokens.Colors.inkFaint)
            HStack(spacing: LiulianTokens.Spacing.s3) {
                LiulianButton("Primary", variant: .primary) {}
                LiulianButton("Secondary", variant: .secondary) {}
                LiulianButton("Ghost", variant: .ghost) {}
                LiulianButton("Danger", variant: .danger) {}
            }

            LiulianText("Sizes", variant: .monoLabel, color: LiulianTokens.Colors.inkFaint)
            HStack(spacing: LiulianTokens.Spacing.s3) {
                LiulianButton("Small", size: .sm) {}
                LiulianButton("Medium", size: .md) {}
                LiulianButton("Large", size: .lg) {}
            }

            LiulianText("States", variant: .monoLabel, color: LiulianTokens.Colors.inkFaint)
            HStack(spacing: LiulianTokens.Spacing.s3) {
                LiulianButton("Loading", loading: true) {}
                LiulianButton("Disabled", disabled: true) {}
            }
        }
        .padding(LiulianTokens.Spacing.s7)
        .background(LiulianTokens.Colors.canvasWarm)
    }
}
#endif

File: android/liulian-ui/src/main/kotlin/io/liulian/ui/LiulianButton.kt
package io.liulian.ui

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.platform.LocalFocusManager
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import io.liulian.tokens.LiulianTokens

enum class ButtonVariant { Primary, Secondary, Ghost, Danger }
enum class ButtonSize { Sm, Md, Lg }

/**
 * LiulianButton — Compose impl per ui-spec/Button.spec.md.
 *
 * Built on Compose Foundation (Box + clickable, NOT Material Button).
 * Disables Material ripple, renders own press scale + focus ring.
 */
@Composable
fun LiulianButton(
    label: String,
    onPress: () -> Unit,
    modifier: Modifier = Modifier,
    variant: ButtonVariant = ButtonVariant.Primary,
    size: ButtonSize = ButtonSize.Md,
    loading: Boolean = false,
    disabled: Boolean = false,
) {
    val interactionSource = remember { MutableInteractionSource() }
    val pressed by interactionSource.collectIsPressedAsState()

    val scale by animateFloatAsState(
        targetValue = if (pressed) LiulianTokens.Control.ScaleActive.value else 1f,
        animationSpec = tween(
            durationMillis = LiulianTokens.Duration.instant,
            easing = LiulianTokens.Easing.easeOut
        ),
        label = "buttonPressScale"
    )

    val height = when (size) {
        ButtonSize.Sm -> LiulianTokens.Control.Height.sm
        ButtonSize.Md -> LiulianTokens.Control.Height.md
        ButtonSize.Lg -> LiulianTokens.Control.Height.lg
    }
    val paddingX = when (size) {
        ButtonSize.Sm -> LiulianTokens.Control.PaddingX.sm
        ButtonSize.Md -> LiulianTokens.Control.PaddingX.md
        ButtonSize.Lg -> LiulianTokens.Control.PaddingX.lg
    }
    val textVariant = when (size) {
        ButtonSize.Sm -> LiulianTextVariant.Caption
        ButtonSize.Md -> LiulianTextVariant.Body
        ButtonSize.Lg -> LiulianTextVariant.Subtitle
    }

    val bg: Color = when (variant) {
        ButtonVariant.Primary -> LiulianTokens.Colors.unibeRed
        ButtonVariant.Secondary -> LiulianTokens.Colors.surfacePure
        ButtonVariant.Ghost -> Color.Transparent
        ButtonVariant.Danger -> LiulianTokens.Colors.unibeRedDeep
    }
    val fg: Color = when (variant) {
        ButtonVariant.Primary, ButtonVariant.Danger -> LiulianTokens.Colors.surfacePure
        ButtonVariant.Secondary -> LiulianTokens.Colors.inkCharcoal
        ButtonVariant.Ghost -> LiulianTokens.Colors.inkMuted
    }

    Box(
        modifier = modifier
            .defaultMinSize(minHeight = LiulianTokens.Touch.minTargetAndroid)
            .graphicsLayer { scaleX = scale; scaleY = scale }
            .alpha(if (disabled) LiulianTokens.Opacity.disabled else 1f)
            .clip(RoundedCornerShape(LiulianTokens.Radius.md))
            .background(bg)
            .let {
                if (variant == ButtonVariant.Secondary) {
                    it.border(
                        width = 1.dp,
                        color = LiulianTokens.Colors.hairline,
                        shape = RoundedCornerShape(LiulianTokens.Radius.md)
                    )
                } else it
            }
            .clickable(
                interactionSource = interactionSource,
                indication = null,
                enabled = !disabled && !loading,
                role = Role.Button,
                onClick = onPress,
            )
            .height(height)
            .padding(horizontal = paddingX),
        contentAlignment = Alignment.Center,
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(LiulianTokens.Control.IconGap.value),
        ) {
            if (loading) {
                // Minimal spinner placeholder; replace with brand spinner when delivered.
                Box(
                    modifier = Modifier
                        .height(LiulianTokens.Control.IconSize.md)
                        .background(fg.copy(alpha = 0.6f), shape = RoundedCornerShape(50))
                )
            } else {
                LiulianText(label, variant = textVariant, color = fg)
            }
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFFFBFBFA)
@Composable
private fun LiulianButtonPreview() {
    Column(
        modifier = Modifier
            .background(LiulianTokens.Colors.canvasWarm)
            .padding(LiulianTokens.Spacing.s7),
        verticalArrangement = Arrangement.spacedBy(LiulianTokens.Spacing.s4),
    ) {
        Row(horizontalArrangement = Arrangement.spacedBy(LiulianTokens.Spacing.s3)) {
            LiulianButton("Primary", onPress = {})
            LiulianButton("Secondary", onPress = {}, variant = ButtonVariant.Secondary)
            LiulianButton("Ghost", onPress = {}, variant = ButtonVariant.Ghost)
            LiulianButton("Danger", onPress = {}, variant = ButtonVariant.Danger)
        }
        Row(horizontalArrangement = Arrangement.spacedBy(LiulianTokens.Spacing.s3)) {
            LiulianButton("Small", onPress = {}, size = ButtonSize.Sm)
            LiulianButton("Medium", onPress = {}, size = ButtonSize.Md)
            LiulianButton("Large", onPress = {}, size = ButtonSize.Lg)
        }
        Row(horizontalArrangement = Arrangement.spacedBy(LiulianTokens.Spacing.s3)) {
            LiulianButton("Loading", onPress = {}, loading = true)
            LiulianButton("Disabled", onPress = {}, disabled = true)
        }
    }
}
